packages/solarwinds-apm/src/processing/transaction-name.ts

```diff
@@ -0,0 +1,132 @@
+
+import { getRootOrEntry, isRootOrEntry } from "./parent-span.js"
+
+const TRANSACTION_NAME_POOL_TTL = 60 * 1000 // 1 minute
+const TRANSACTION_NAME_POOL_MAX = 200
+const TRANSACTION_NAME_DEFAULT = "other"
+const TRANSACTION_NAME_ATTRIBUTE = "sw.transaction"
+
+/**
+ * Sets a custom name for the current transaction
+ *
+ * @param name - Custom transaction name
+ * @returns Whether the name was successfully set
+ */
+export function setTransactionName(name: string): boolean {
+  const active = trace.getActiveSpan()
+  const rootOrEntry = active && getRootOrEntry(active)
+  if (!rootOrEntry) {
+    return false
+  }
+
+  rootOrEntry.setAttribute(TRANSACTION_NAME_ATTRIBUTE, name)
+  return true
+}
+
+/** Processor that sets the transaction name attribute on spans */
+export class TransactionNameProcessor
+  extends NoopSpanProcessor
+  implements SpanProcessor
+{
+  readonly #pool = new TransactionNamePool({
+    max: TRANSACTION_NAME_POOL_MAX,
+    ttl: TRANSACTION_NAME_POOL_TTL,
+    default: TRANSACTION_NAME_DEFAULT,
+  })
+  readonly #defaultName?: string
+
+  constructor(config: SwConfiguration) {
+    super()
+    this.#defaultName = config.transactionName
+  }
+
+  override onEnd(span: ReadableSpan): void {
+    if (!isRootOrEntry(span)) {
+      return
+    }
+
+    let name = span.attributes[TRANSACTION_NAME_ATTRIBUTE]
+    if (typeof name !== "string") {
+      name = this.#defaultName ?? computedTransactionName(span)
+    }
+    name = this.#pool.registered(name)
+
+    span.attributes[TRANSACTION_NAME_ATTRIBUTE] = name
+  }
+}
+
+/** Computes a transaction name from a span and its attributes */
+export function computedTransactionName(span: ReadableSpan): string {
+  if (typeof process.env.AWS_LAMBDA_FUNCTION_NAME === "string") {
+    return process.env.AWS_LAMBDA_FUNCTION_NAME
+  } else if (typeof span.attributes[SEMATTRS_HTTP_ROUTE] === "string") {
+    return span.attributes[SEMATTRS_HTTP_ROUTE]
+  } else if (typeof span.attributes[SEMATTRS_HTTP_TARGET] === "string") {
+    // split on slashes and keep the first 3 segments
+    // where the first segment is an empty string before the first slash
+    return span.attributes[SEMATTRS_HTTP_TARGET].split("/", 3).join("/")
+  } else {
+    return span.name
+  }
+}
+
+/** A pool that prevents explosion of cardinality in transaction names */
+export class TransactionNamePool {
+  readonly #pool = new Map<string, NodeJS.Timeout>()
+
+  readonly #max: number
+  readonly #ttl: number
+  readonly #default: string
+
+  constructor(options: { max: number; ttl: number; default: string }) {
+    this.#max = options.max
+    this.#ttl = options.ttl
+    this.#default = options.default
+  }
+
+  /** Given a desired transaction name return the one that should be used */
+  registered(name: string): string {
+    const existing = this.#pool.get(name)
+    if (existing !== undefined) {
+      clearTimeout(existing)
+    }
+
+    if (existing !== undefined || this.#pool.size < this.#max) {
```

Comment:
Purpose of this pool makes sense, but I don't think I understand how it works. Is it this?: processor has calculated a name to use (custom or from lambda/http/span name) then checks it against this pool. If the name exists in the pool and we're not yet at pool max, ??? delete after a certain amount of time?

Reply:
Yeah, check if either the pool already had the name or if there's some room we use it, and schedule it to be removed from the pool after the TTL to make some free room. We have to store the scheduled removal tasks themselves in the pool to be able to cancel them when a name is reused.

New name and there is room -> Add the name to pool and schedule to be removed after TTL -> Return name
New name but there is no room -> Return default name
Existing name -> Cancel the removal previously scheduled -> Schedule a new removal -> Return name

Reply:
Adding a comment to clarify that in the implementation !

Reply:
Thank you that's great!